### ISSUE TITLE
🐛 Bug Report: search error Collating documents for software-catalog failed: FetchError: request to https://localhost:7777/api/catalog/

### ISSUE THREAD

**Issue by anishaanil1** [backstage/backstage#20518]
### 📜 Description

configured backstage to use HTTPS mode and used the below document --> https://github.com/backstage/backstage/blob/6b466acbadf30503cb35469f1a78d40e76b7f6fa/contrib/docs/tutorials/enable-ssl-self-signed.md?plain=1#L34

Post this configuration and starting the application it fails with the below error

```
Duration":"PT15M"} type=taskManager task=search_index_techdocs
2023-10-10T16:03:26.777Z search info Collating documents for software-catalog via DefaultCatalogCollatorFactory type=plugin documentType=software-catalog
2023-10-10T16:03:26.787Z search info Collating documents for techdocs via DefaultTechDocsCollatorFactory type=plugin documentType=techdocs
2023-10-10T16:03:26.796Z search warn Index for software-catalog was not created: an error was encountered type=plugin documentType=software-catalog
2023-10-10T16:03:26.796Z search error Collating documents for software-catalog failed: FetchError: request to https://localhost:7777/api/catalog/entities?offset=0&limit=500 failed, reason: unable to verify the first certificate type=plugin documentType=software-catalog
2023-10-10T16:03:26.797Z backstage error request to https://localhost:7777/api/catalog/entities?offset=0&limit=500 failed, reason: unable to verify the first certificate type=system task=search_index_software_catalog errno=UNABLE_TO_VERIFY_LEAF_SIGNATURE code=UNABLE_TO_VERIFY_LEAF_SIGNATURE stack=FetchError: request to https://localhost:7777/api/catalog/entities?offset=0&limit=500 failed, reason: unable to verify the first certificate
    at ClientRequest.<anonymous> (/data/test_backstage/backstage1/backstage/node_modules/node-fetch/lib/index.js:1501:11)
    at ClientRequest.emit (node:events:514:28)
    at TLSSocket.socketErrorListener (node:_http_client:501:9)
    at TLSSocket.emit (node:events:514:28)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
2023-10-10T16:03:26.799Z search warn Index for techdocs was not created: an error was encountered type=plugin documentType=techdocs
```

This is definitely not an error with certificates as the backend should start with the hostname but somehow it is automatically converted to localhost which then fails with the certs.

If i trigger the api directly with the actual hostname in the certificate it works without any errors. At this point I'm unsure how the backend for docs is converting automatically to local host

### 👍 Expected behavior

It should point to the actual value of the hostname defined in the app-config.yaml

### 👎 Actual Behavior with Screenshots

It is pointing to localhost and not picking values from app-config.yaml file

### 👟 Reproduction steps

1> Create a new instance of Backstage
2> Change the instance from http to https by generating a new certificate with the steps mentioned in the document
3> Start the backend server and the issue will be reproduced
4> The issue is seen even when you configure it with kubernetes and use ingress in https mode to open backstage

### 📃 Provide the context for the Bug.

Without having it in https mode we cannot authenticate using our Azure ADFS and will not be able to do any  RBAC for users.
The solution will be deemed as not helpful if we cannot setup the authentication

### 🖥️ Your Environment

```
yarn backstage-cli info
yarn run v1.22.19
$ /data/test_backstage/backstage1/backstage/node_modules/.bin/backstage-cli info
OS:   Linux 5.15.0-1044-aws - linux/x64
node: v18.17.0
yarn: 1.22.19
cli:  0.22.13 (installed)
backstage:  1.18.0

Dependencies:
  @backstage/app-defaults                                          1.4.3
  @backstage/backend-app-api                                       0.5.5
  @backstage/backend-common                                        0.19.7
  @backstage/backend-dev-utils                                     0.1.1
  @backstage/backend-openapi-utils                                 0.0.4
  @backstage/backend-plugin-api                                    0.6.5
  @backstage/backend-tasks                                         0.5.10
  @backstage/catalog-client                                        1.4.4
  @backstage/catalog-model                                         1.4.2
  @backstage/cli-common                                            0.1.12
  @backstage/cli-node                                              0.1.4
  @backstage/cli                                                   0.22.13
  @backstage/config-loader                                         1.5.0
  @backstage/config                                                1.1.0
  @backstage/core-app-api                                          1.10.0
  @backstage/core-components                                       0.13.5
  @backstage/core-plugin-api                                       1.6.0
  @backstage/errors                                                1.2.2
  @backstage/eslint-plugin                                         0.1.3
  @backstage/integration-aws-node                                  0.1.6
  @backstage/integration-react                                     1.1.19
  @backstage/integration                                           1.7.0
  @backstage/plugin-api-docs                                       0.9.11
  @backstage/plugin-app-backend                                    0.3.53
  @backstage/plugin-app-node                                       0.1.5
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.1.2
  @backstage/plugin-auth-backend-module-github-provider            0.1.2
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.2
  @backstage/plugin-auth-backend-module-google-provider            0.1.2
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.2
  @backstage/plugin-auth-backend                                   0.19.2
  @backstage/plugin-auth-node                                      0.3.2
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.2
  @backstage/plugin-catalog-backend                                1.13.3
  @backstage/plugin-catalog-common                                 1.0.16
  @backstage/plugin-catalog-graph                                  0.2.36
  @backstage/plugin-catalog-import                                 0.10.0
  @backstage/plugin-catalog-node                                   1.4.6
  @backstage/plugin-catalog-react                                  1.8.4
  @backstage/plugin-catalog                                        1.13.0
  @backstage/plugin-events-node                                    0.2.14
  @backstage/plugin-github-actions                                 0.6.5
  @backstage/plugin-org                                            0.6.14
  @backstage/plugin-permission-common                              0.7.8
  @backstage/plugin-permission-node                                0.7.16
  @backstage/plugin-permission-react                               0.4.15
  @backstage/plugin-proxy-backend                                  0.4.2
  @backstage/plugin-scaffolder-backend                             1.17.3
  @backstage/plugin-scaffolder-common                              1.4.1
  @backstage/plugin-scaffolder-node                                0.2.5
  @backstage/plugin-scaffolder-react                               1.5.5
  @backstage/plugin-scaffolder                                     1.15.0
  @backstage/plugin-search-backend-module-catalog                  0.1.9
  @backstage/plugin-search-backend-module-pg                       0.5.14
  @backstage/plugin-search-backend-module-techdocs                 0.1.9
  @backstage/plugin-search-backend-node                            1.2.9
  @backstage/plugin-search-backend                                 1.4.5
  @backstage/plugin-search-common                                  1.2.6
  @backstage/plugin-search-react                                   1.7.0
  @backstage/plugin-search                                         1.4.0
  @backstage/plugin-tech-radar                                     0.6.8
  @backstage/plugin-techdocs-backend                               1.7.2
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.0
  @backstage/plugin-techdocs-node                                  1.8.2
  @backstage/plugin-techdocs-react                                 1.1.11
  @backstage/plugin-techdocs                                       1.7.0
  @backstage/plugin-user-settings                                  0.7.10
  @backstage/release-manifests                                     0.0.10
  @backstage/test-utils                                            1.4.3
  @backstage/theme                                                 0.4.2
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.5
Done in 0.68s.
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by camilaibs** [comment]
Hi @anishaanil1, thanks for reaching out to us and also being happy to contribute with a fix.
Just to make sure we know how to exactly reproduce the steps to simulate the error, could we answer the following questions?
- What is the value of `backend.baseUrl` on your `app-config.yaml` file?
- Are you passing a custom catalog client for the `DefaultCatalogCollatorFactory` in the backend application?

**Comment by anishaanil1** [comment]
Please find the answers below. 

All i did was just install backstage as default. --> npx @backstage/create-app@latest
Then checked if the system was up and running and it was all good. 

Next use the link to change the http to https for both front end and backend.
https://github.com/backstage/backstage/blob/6b466acbadf30503cb35469f1a78d40e76b7f6fa/contrib/docs/tutorials/enable-ssl-self-signed.md?plain=1#L34

Then modified the file: Please find the complete app-config.yaml file

```
app:
  title: Scaffolded Backstage App
  baseUrl: https://backstage.test.com:3333


organization:
  name: My Company

backend:
  # Used for enabling authentication, secret is shared by all backend plugins
  # See https://backstage.io/docs/auth/service-to-service-auth for
  # information on the format
  # auth:
  #   keys:
  #     - secret: ${BACKEND_SECRET}
  baseUrl: https://backstage.test.com:7777
  listen:
    port: 7777
  https:
    certificate:
      cert: |
        -----BEGIN CERTIFICATE-----
        MIIEkTCCAvmgAwIBAgIRAN1sDHfBOrICy1mtFj97N7gwDQYJKoZIhvcNAQELBQAw
        gYExHjAcBgNVBAoTFW1rY2VydCBkZXZlbG9wbWVudCBDQTErMCkGA1UECwwicm9v
                OIN62gr+bmWNy7Z5U/lVAS83wutv0iVG03Jn2GEYxqWMSOI4N31C5A3vmN8T7W6X
        7EyirikN7cRPZxRaF6i2bDiwISG87+RBNEsLHFcZypbz3uDCB/njGnETzTywQCch
        E2tfJSPXSBxZM8Bd7cyznegEFhNm
        -----END CERTIFICATE-----
      key:
        $file: ./certs1/backstage.test.com.com+4-key.pem
    # Uncomment the following host directive to bind to specific interfaces
    # host: 127.0.0.1
  csp:
    connect-src: ["'self'", 'http:', 'https:']
    upgrade-insecure-requests: false
    # Content-Security-Policy directives follow the Helmet format: https://helmetjs.github.io/#reference
    # Default Helmet Content-Security-Policy values can be removed by setting the key to false
  cors:
    origin: https://backstage.test.com:3333
    methods: [GET, HEAD, PATCH, POST, PUT, DELETE]
    credentials: true
  # This is for local development only, it is not recommended to use this in production
  # The production database configuration is stored in app-config.production.yaml
  database:
    client: better-sqlite3
    connection: ':memory:'
  # workingDirectory: /tmp # Use this to configure a working directory for the scaffolder, defaults to the OS temp-dir

integrations:
  github:
    - host: github.com
      # This is a Personal Access Token or PAT from GitHub. You can find out how to generate this token, and more information
      # about setting up the GitHub integration here: https://backstage.io/docs/getting-started/configuration#setting-up-a-github-integration
      token: ${GITHUB_TOKEN}
    ### Example for how to add your GitHub Enterprise instance using the API:
    # - host: ghe.example.net
    #   apiBaseUrl: https://ghe.example.net/api/v3
    #   token: ${GHE_TOKEN}

proxy:
  ### Example for how to add a proxy endpoint for the frontend.
  ### A typical reason to do this is to handle HTTPS and CORS for internal services.
  # endpoints:
  #   '/test':
  #     target: 'https://example.com'
  #     changeOrigin: true

# Reference documentation http://backstage.io/docs/features/techdocs/configuration
# Note: After experimenting with basic setup, use CI/CD to generate docs
# and an external cloud storage when deploying TechDocs for production use-case.
# https://backstage.io/docs/features/techdocs/how-to-guides#how-to-migrate-from-techdocs-basic-to-recommended-deployment-approach
techdocs:
  builder: 'local' # Alternatives - 'external'
  generator:
    runIn: 'docker' # Alternatives - 'local'
  publisher:
    type: 'local' # Alternatives - 'googleGcs' or 'awsS3'. Read documentation for using alternatives.

auth:
  # see https://backstage.io/docs/auth/ to learn about auth providers
  providers: {}

scaffolder:
  # see https://backstage.io/docs/features/software-templates/configuration for software template options

catalog:
  import:
    entityFilename: catalog-info.yaml
    pullRequestBranchName: backstage-integration
  rules:
    - allow: [Component, System, API, Resource, Location]
  locations:
    # Local example data, file locations are relative to the backend process, typically `packages/backend`
    - type: file
      target: ../../examples/entities.yaml

    # Local example template
    - type: file
      target: ../../examples/template/template.yaml
      rules:
        - allow: [Template]

    # Local example organizational data
    - type: file
      target: ../../examples/org.yaml
      rules:
        - allow: [User, Group]

    ## Uncomment these lines to add more example data
    # - type: url
    #   target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all.yaml

    ## Uncomment these lines to add an example org
    # - type: url
    #   target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/acme-corp.yaml
    #   rules:
    #     - allow: [User, Group]
```

Are you passing a custom catalog client for the DefaultCatalogCollatorFactory in the backend application?
No. I'm keeping everything as defaults as our first requirement is to get this instance converted to https.

Addional testing:
Kept the app-config.yaml file as default  and then added a nginx as reverse proxy and it also shows the same error

**Comment by anishaanil1** [comment]
Any Updates?????? We are in process of finalizing the Software Asset tool and if it cannot be configured in https mode........it becomes a big bottle neck

**Comment by freben** [comment]
You inlined the cert in yaml. Not sure if that can interfere due to whitespace or similar. Would be useful to try to use $file like for the private one. Use absolute paths, to ensure that they can be found no matter how the backend is packaged / started.

The request that happens is made backend-to-backend. So it's important to note that the backend that has the search plugin tried to make a request to a backend that has the catalog plugin. Those may or may not be co-hosted of course. But it's important that both those hosts have the same https config AND are launched with the `NODE_EXTRA_CA_CERTS` env var to make sure that trust is established properly. If the base URL for example pointed to an ingress that is meant for browsers to talk to but isn't "curlable", then things would also fail.

When you made the experiment with the reverse proxy, did you change the backend baseUrl to point to that proxy? And did you make it so that the backend does NOT have https set up then?

The dream scenario is:

```
browser ------> proxy ------> backend1 ------> backend2
         https         http             http
```

In that situation, the entire problem space of https is left out of Node's (and Backstage's) responsibility, apart from one thing: in the backend, you make a specific `discoveryApi` implementation that returns base URLs pointing directly to backends, not the proxy.

**Comment by anishaanil1** [comment]
i have tried all those combinations with Nginx backend and withourt..........but still it fails with the same error. Are you not able to replicate the behavior?????
If it is working on your system............Can you please let me know how exactly have you configured the system.

**Comment by emmaindal** [comment]
Hello @anishaanil1, just wanted to let you know that I've walked through the steps you provided to reproduce the issue. And I'm also running into a similar error just running the backend locally. 

However I don't see these issues for just search, its for all requests to the catalog api in my case. Will have to get input from @backstage/maintainers here around the steps for the certification. My app-config looks like the following

```
backend:
  baseUrl: https://localhost:7007
  https:
    certificate:
      key:
        $file: ./localhost-key.pem
      cert:
        $file: ./localhost.pem
```

were running the backend with (ofc pointed to my file): `NODE_EXTRA_CA_CERTS=/absolute/path/to/cert.pem yarn start-backend`

@freben any recommendations here?

**Comment by anishaanil1** [comment]
@freben we are completely stalled in our organization as we are unable to set it up with HTTPS.
Without HTTPS we cannot integrate with LDAP from Azure as the callback url needs it to be in https format.

We tried nginx aswell on front end and it still fails with the same error.

Any directions is greatly appreciated

**Comment by anishaanil1** [comment]
@freben We have now configured it with f5 Load balancer ad every time we hit the https://backstage.test.com it gives me the --> Invalid Host header. The entire system is working with http mode in the backend and supported by NGINX or F5 in the front end
 This is really blocking our production deployment..........Any directions is greatly appreciated

**Comment by freben** [comment]
`backend.baseUrl` is the endpoint that the browser should use when it's trying to reach out to the backend. I suspect that should be set to `https://<loadbalancer>` at the standard HTTPS port instead of 7777, right?

Same thing then for `app.baseUrl`, likely you want that to also be set to `https://<loadbalancer>`.

Am I correct in understanding that this is a TLS terminating load balancer? If so, I presume that you want to configure it to forward traffic to Backstage as HTTP, not HTTPS, vastly simplifying things.

Then `backend.listen.port` sets that the listening socket port of the backend. Your load balancer should be set to forward traffic to `http://<service>:7777` if this is what you've set as the listen port. Note http, not https, in that example.

**Comment by anishaanil1** [comment]
@freben , @emmaindal , We have made the following changes and now the error points to plugin. I don't see a location to update the plugin with https certs. Where exactly should the plugin SSL be updated

Configuration in App-config.yaml

******************
vi app-config.yaml
app:
  title: Scaffolded Backstage App
  baseUrl: http://techshelf.test.com:3333
  https:
    certificate:
      cert:
        $file: /etc/pki/tls/certs/techshelf.test.com_with_chain.crt
      key:
        $file: /etc/pki/tls/private/techshelf.test.com.key

organization:
  name: My Company

backend:
  # Used for enabling authentication, secret is shared by all backend plugins
  # See https://backstage.io/docs/auth/service-to-service-auth for
  # information on the format
  # auth:
  #   keys:
  #     - secret: ${BACKEND_SECRET}
  baseUrl: https://techshelf.test.com:7777
  listen:
    port: 7777
  https:
    certificate:
      cert:
        $file: /etc/pki/tls/certs/techshelf.test.com_with_chain.crt
      key:
        $file: /etc/pki/tls/private/techshelf.test.com.key
    # Uncomment the following host directive to bind to specific interfaces
    # host: 127.0.0.1
  csp:
    connect-src: ["'self'", 'http:', 'https:']
    # Content-Security-Policy directives follow the Helmet format: https://helmetjs.github.io/#reference
    # Default Helmet Content-Security-Policy values can be removed by setting the key to false
  cors:
    origin: https://techshelf.test.com:3333
    methods: [GET, HEAD, PATCH, POST, PUT, DELETE]
    credentials: true

******************
Configuration in Nginx ssl.conf
******************
server {
    listen 80 default_server;
    listen [::]:80 default_server;

    location / {
        #proxy_pass http://techshelf.test.com:3000;
        return 301 https://$host$request_uri;
    }
}

server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    ssl_certificate /etc/pki/tls/certs/techshelf.test.com_with_chain.crt;
    ssl_certificate_key /etc/pki/tls/private/techshelf.test.com.key;
    ssl_session_timeout 1d;
    ssl_session_cache shared:MozSSL:10m;  # about 40000 sessions
    ssl_session_tickets off;

    root /usr/share/nginx/html;

    location / {
     proxy_set_header X-Forwarded-For $remote_addr;
     proxy_set_header Host $http_host;
     proxy_pass http://techshelf.test.com:3333;
        #return 301 https://$host$request_uri;
    }

    # curl https://ssl-config.mozilla.org/ffdhe2048.txt > /path/to/dhparam
    ssl_dhparam /etc/nginx/conf.d/dhparam;

******************
Error:
******************
[1] 2023-11-27T08:24:28.437Z search info Collating documents for software-catalog via DefaultCatalogCollatorFactory type=plugin documentType=software-catalog
[1] 2023-11-27T08:24:28.441Z search info Collating documents for techdocs via DefaultTechDocsCollatorFactory type=plugin documentType=techdocs
[1] 2023-11-27T08:24:28.456Z search warn Index for software-catalog was not created: an error was encountered type=plugin documentType=software-catalog
[1] 2023-11-27T08:24:28.456Z search error Collating documents for software-catalog failed: FetchError: request to https://localhost:7777/api/catalog/entities?offset=0&limit=500 failed, reason: unable to verify the first certificate type=plugin documentType=software-catalog
[1] 2023-11-27T08:24:28.456Z backstage error request to https://localhost:7777/api/catalog/entities?offset=0&limit=500 failed, reason: unable to verify the first certificate type=system task=search_index_software_catalog errno=UNABLE_TO_VERIFY_LEAF_SIGNATURE code=UNABLE_TO_VERIFY_LEAF_SIGNATURE stack=FetchError: request to https://localhost:7777/api/catalog/entities?offset=0&limit=500 failed, reason: unable to verify the first certificate
[1]     at ClientRequest.<anonymous> (/data/https/backstage/node_modules/node-fetch/lib/index.js:1501:11)
[1]     at ClientRequest.emit (node:events:517:28)
[1]     at ClientRequest.emit (node:domain:489:12)
[1]     at TLSSocket.socketErrorListener (node:_http_client:501:9)
[1]     at TLSSocket.emit (node:events:517:28)
[1]     at TLSSocket.emit (node:domain:489:12)
[1]     at emitErrorNT (node:internal/streams/destroy:151:8)
[1]     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
[1]     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
[1] 2023-11-27T08:24:28.462Z search warn Index for techdocs was not created: an error was encountered type=plugin documentType=techdocs
[1] 2023-11-27T08:24:28.463Z search error Collating documents for techdocs failed: FetchError: request to https://localhost:7777/api/catalog/entities?filter=metadata.annotations.backstage.io%2Ftechdocs-ref&offset=0&limit=500 failed, reason: unable to verify the first certificate type=plugin documentType=techdocs
[1] 2023-11-27T08:24:28.463Z backstage error request to https://localhost:7777/api/catalog/entities?filter=metadata.annotations.backstage.io%2Ftechdocs-ref&offset=0&limit=500 failed, reason: unable to verify the first certificate type=system task=search_index_techdocs errno=UNABLE_TO_VERIFY_LEAF_SIGNATURE code=UNABLE_TO_VERIFY_LEAF_SIGNATURE stack=FetchError: request to https://localhost:7777/api/catalog/entities?filter=metadata.annotations.backstage.io%2Ftechdocs-ref&offset=0&limit=500 failed, reason: unable to verify the first certificate
[1]     at ClientRequest.<anonymous> (/data/https/backstage/node_modules/node-fetch/lib/index.js:1501:11)
[1]     at ClientRequest.emit (node:events:517:28)
[1]     at ClientRequest.emit (node:domain:489:12)
[1]     at TLSSocket.socketErrorListener (node:_http_client:501:9)
[1]     at TLSSocket.emit (node:events:517:28)
[1]     at TLSSocket.emit (node:domain:489:12)
[1]     at emitErrorNT (node:internal/streams/destroy:151:8)
[1]     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
[1]     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)


******************
This doesn't looks like a certs issue as we are using this for other application aswell

**Comment by anishaanil1** [comment]
@freben , @emmaindal , Any thoughts. 
I also see that the "https://demo.backstage.io/" is in https mode..............What architecture is it using????? Are you running it on containers or k8s.............Any leads as we unable to push this setup to production without having it in HTTPS format

**Comment by emmaindal** [comment]
Hello @anishaanil1 from the logs what I can see is 

```
backstage error request to https://localhost:7777/api/catalog/entities?offset=0&limit=500 failed, reason: unable to verify the first certificate type=system task=search_index_software_catalog errno=UNABLE_TO_VERIFY_LEAF_SIGNATURE code=UNABLE_TO_VERIFY_LEAF_SIGNATURE stack=FetchError: request to https://localhost:7777/api/catalog/entities?offset=0&limit=500 failed, reason: unable to verify the first certificate
```
 
 this to me implies two things: 
 1. that you are running against a local backend (is this really what you want)? 
 2. that the problem is in the request to the catalog as its unable to verify the certificate 
 
but for me to help you continue investigate if this is a search related bug can I ask you to comment out collators that depend on the catalog in your backend search setup. E.g only keep the `ToolDocumentCollatorFactory` out of the following 
 
```ts
    /* 
   indexBuilder.addCollator({
    schedule,
    factory: DefaultCatalogCollatorFactory.fromConfig(env.config, {
      discovery: env.discovery,
      tokenManager: env.tokenManager,
    }),
  });

  indexBuilder.addCollator({
    schedule,
    factory: DefaultTechDocsCollatorFactory.fromConfig(env.config, {
      discovery: env.discovery,
      logger: env.logger,
      tokenManager: env.tokenManager,
    }),
  });
  */ 
  
  indexBuilder.addCollator({
    schedule,
    factory: ToolDocumentCollatorFactory.fromConfig(env.config, {
      discovery: env.discovery,
      logger: env.logger,
    }),
  });
```  

and run your app again. Do you still see the same errors?

**Comment by anishaanil1** [comment]
1. that you are running against a local backend (is this really what you want)?
Is there another way you want me to set this up.............I followed the documentation to start with something really simple......and ended up with the certs issue.

https://backstage.io/docs/getting-started/
https://backstage.io/docs/getting-started/create-an-app/

2. that the problem is in the request to the catalog as its unable to verify the certificate
We have setup everything as default. 
Step-1
1> Install all prerequisites
2> npx @backstage/create-app@latest
3> Configure the postgress database on the same machine  
4> yarn Dev

--> Which file do you want to edit. Can you please provide the path.

**Comment by emmaindal** [comment]
it should most likely be the `packages/backend/src/plugins/search.ts` file

**Comment by emmaindal** [comment]
regarding the cert verification errors you see, I can see a similar issue opened here https://github.com/backstage/backstage/issues/3399 with a couple of different suggestions. Can you see if this helps you resolve the problem?

**Comment by freben** [comment]
The ReferenceError in your previous comment clearly indicates that you have an actual code error that you'll have to fix. Are you using an editor that helps with syntax highlighting and linting and completion etc?

**Comment by freben** [comment]
Your config seems confused. You set `app.baseUrl` to have an `http` protocol but didn't you say that you were terminating it as https in nginx? And you say that it's on port 3333 even though the nginx conf uses the standard 443 port. The `backend.baseUrl` points to port 7777, does the nginx proxy listen there? Shouldn't that one also be just `https://techshelf.test.com`? Your `cors.origin` also has the unexpected port number there; this should all be as seen from the browser's point of view and it's gonna be just `https://techshelf.test.com` in the location bar of the browser right?

Also you've set your nginx to talk to `http://techshelf.test.com:3333` which is NOT https, but you've enabled https on the backstage instance... this is all pretty confused.

**Comment by freben** [comment]
To be more concrete, if you're terminating TLS with nginx, I'd expect something like the following. Bear with me because it's late and I'm typing this freehand without testing it. Also, note that this ASSUMES that `https://techshelf.test.com` is what the actual address bar in the browser is expected to show.

```yaml
app:
  baseUrl: https://techshelf.test.com
backend:
  baseUrl: https://techshelf.test.com
  listen:
    port: 7777
  cors:
    origin: https://techshelf.test.com
```

Note that this is all https URLs, but the server DOES NOT enable https on the 7777 port because we expect to get http connections on that port, not https.

Then the nginx would do

```
proxy_pass http://techshelf.test.com:7777
```

NOTE: Now when the backend plugins makes requests to other backend plugins, they will ALSO use the `backend.baseUrl`. This means that node makes hairpinning HTTPS requests to your nginx proxy, which then forwards them back again. That's typically perfectly fine, BUT you will need to start the backend process with `NODE_EXTRA_CA_CERTS=/absolute/path/to/cert.pem` so that it can recognize the cert that nginx presents to it.

The other alternative is to supply a custom duscovery implementation [in your backend setup](https://github.com/backstage/backstage/blob/9875cfb30b7ca5ae2047472201337b30796068fb/packages/backend/src/index.ts#L87) that always resolves to `http://localhost:7777/api/<plugin-id>`. Then you don't worry about https at all, but lose out on the possible load balancing / failover / whatnot benefits that the previous method could give.

**Comment by anishaanil1** [comment]
It doesn't seem to be working and failing with the same error. 
```
app:
  title: Scaffolded Backstage App
  baseUrl: https://techshelf.test.com
    #baseUrl: https://techshelf.test.com
  https:
    certificate:
      cert:
        $file: /etc/pki/tls/certs/techshelf.test.com_with_chain.crt
      key:
        $file: /etc/pki/tls/private/techshelf.test.com.key

backend:
  # Used for enabling authentication, secret is shared by all backend plugins
  # See https://backstage.io/docs/auth/service-to-service-auth for
  # information on the format
  # auth:
  #   keys:
  #     - secret: ${BACKEND_SECRET}
  #baseUrl: https://techshelf.test.com:7777
  baseUrl: https://techshelf.test.com:7777
  listen:
    port: 7777
  https:
    certificate:
      cert:
        $file: /etc/pki/tls/certs/techshelf.test.com_with_chain.crt
      key:
        $file: /etc/pki/tls/private/techshelf.test.com.key

cors:
    origin: https://techshelf.test.com

```

Even before we start nginx it is failed with the same error. If nginx is running in the backend it fails with port 443 conflict.

**Comment by anishaanil1** [comment]
Can't we bypass nginx tself

**Comment by emmaindal** [comment]
Hello @anishaanil1 
I went through the steps you have listed to reproduce this issue one more time and again I am able to reproduce what you see. 

I'll add the steps I took here in case you want to follow them too.

1. created new app using instructions in https://backstage.io/docs/getting-started/create-an-app
2. were running the newly created app locally with `yarn dev` to make sure it was running without issues.
3. followed the [Enabling SSL for Local Testing guide]  (https://github.com/backstage/backstage/blob/a898b29ba5927edcff6915ec1a653dde3d9b6384/contrib/docs/tutorials/enable-ssl-self-signed.md).
4. were running the newly created app again with `NODE_EXTRA_CA_CERTS=/absolute/path/to/cert.pem yarn dev`
5. Got the errors you see report above
6. Disabled search collators in`search.ts` (`DefaultTechDocsCollatorFactory` and `DefaultCatalogCollatorFactory`) that communicates with the catalog. 
7. Enabled another collator that do not communicate with catalog to investigate the issue

```bash
yarn add -W @backstage/plugin-search-backend-module-stack-overflow-collator
```

in search.ts
```
import { StackOverflowQuestionsCollatorFactory } from '@backstage/plugin-search-backend-module-stack-overflow-collator';

indexBuilder.addCollator({
    schedule,
    factory: StackOverflowQuestionsCollatorFactory.fromConfig(env.config, {
      logger: env.logger,
      requestParams: {
        tags: ['backstage'],
        site: 'stackoverflow',
        pagesize: 100,
      },
    }),
  });
```

8. got search results indexed successfully
9. enabled a collator that do communicate with other backstage apis (ToolDocumentCollatorFactory)
10. reached the same error again
11. were running the app with `NODE_TLS_REJECT_UNAUTHORIZED=0` using the following command `NODE_EXTRA_CA_CERTS=/absolute/path/to/cert.pem NODE_TLS_REJECT_UNAUTHORIZED=0 yarn dev` I was able to index everything successfully again with all collators enabled (`StackOverflowQuestionsCollatorFactory`, `ToolDocumentCollatorFactory`, `DefaultTechDocsCollatorFactory`, `DefaultCatalogCollatorFactory`)

Of course this means we disabling the certificate validation, which to me sounds bad. And we would like to keep the certificate verification on. 

13. After some additional investigation I found https://github.com/FiloSottile/mkcert/issues/214 which creates a full chain cert file for me and then I could run the app again with `NODE_EXTRA_CA_CERTS=absolute/path/to/fullchain.pem yarn dev`
14. Successful indexing process


With this said, it does actually not seem to be a search bug, so I'll close this issue. But if you disagree and do not get it to work with my help above - please re-open the issue again!

**Comment by kentwongg** [comment]
Found this thread googling an error I got. Read through most of the comments.

I don't think this is a search bug, but it seems to be an issue with something inside backstage.

I have backstage setup on a cloud box and behind a proxy. I gave it the cert files and installed it on the hosts. Getting an error where when I try to auth through Gitlab, I get an error about certs but the URL is troubling:
`
Login failed; caused by FetchError: request to https://localhost:7007/api/catalog/entities/by-name/User/default/kent.wong failed, reason: unable to verify the first certificate`

I don't have localhost anywhere in my app-config and I'm running it in production mode as a docker container. It should at the very least be the DNS value I have as the baseURL, i.e. https://acme.org and not localhost.
